// @gelatonetwork/limit-orders-lib/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = exports.utils = exports.isEthereumChain = exports.GelatoStopLimitOrders = exports.isValidChainIdAndHandler = exports.GelatoLimitOrders = void 0;
const tslib_1 = require("tslib");
var limit_orders_1 = require("./limit-orders");
Object.defineProperty(exports, "GelatoLimitOrders", { enumerable: true, get: function () { return limit_orders_1.GelatoLimitOrders; } });
Object.defineProperty(exports, "isValidChainIdAndHandler", { enumerable: true, get: function () { return limit_orders_1.isValidChainIdAndHandler; } });
var stoplimit_orders_1 = require("./stoplimit-orders");
Object.defineProperty(exports, "GelatoStopLimitOrders", { enumerable: true, get: function () { return stoplimit_orders_1.GelatoStopLimitOrders; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "isEthereumChain", { enumerable: true, get: function () { return utils_1.isEthereumChain; } });
(0, tslib_1.__exportStar)(require("./types"), exports);
exports.utils = (0, tslib_1.__importStar)(require("./utils"));
exports.constants = (0, tslib_1.__importStar)(require("./constants"));
